exonum/src/api_ng/mod.rs

```diff
@@ -0,0 +1,240 @@
+// Copyright 2018 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! API and corresponding utilities.
+
+pub use self::error::Error;
+pub use self::state::ServiceApiState;
+pub use self::with::{FutureResult, Immutable, Mutable, NamedWith, Result, With};
+
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+
+use std::collections::BTreeMap;
+
+use self::backends::actix;
+use blockchain::{Blockchain, SharedNodeState};
+
+pub mod backends;
+pub mod error;
+pub mod node;
+mod state;
+mod with;
+
+/// Trait defines object that could be used as an API backend.
+pub trait ServiceApiBackend: Sized {
+    /// Concrete endpoint handler in the backend.
+    type Handler;
+    /// Concrete output API scope.
+    type Scope;
+
+    /// Adds the given endpoint handler to the backend.
+    fn endpoint<Q, I, R, F, E>(&mut self, name: &'static str, endpoint: E) -> &mut Self
```

Comment:
Well, I don't like &'static requirement here. I think in some situation user may want to generate endpoints on the fly